src/components/views/dialogs/DevtoolsDialog.js

```diff
@@ -0,0 +1,294 @@
+                </div>
+
+                <div className="mx_TextInputDialog_label">
+                    <label htmlFor="evContent"> { _t('Event Content') } </label>
+                </div>
+                <div>
+                    <textarea id="evContent" ref="evContent" className="mx_TextInputDialog_input" defaultValue={"{\n\n}"} cols="63" rows="5" />
+                </div>
+            </div>
+            {this._buttons()}
+        </div>;
+    }
+}
+
+class SendCustomStateEvent extends SendCustomEvent {
+    send(content) {
+        return MatrixClientPeg.get().sendStateEvent(this.props.roomId, this.refs.eventType.value, content,
+            this.refs.stateKey.value);
+    }
+
+    render() {
+        if (this.state.message) {
+            return <div>
+                <div className="mx_Dialog_content">
+                    {this.state.message}
+                </div>
+                {this._buttons()}
+            </div>;
+        }
+
+        return <div>
+            <div className="mx_Dialog_content">
+                <div className="mx_TextInputDialog_label">
+                    <label htmlFor="stateKey"> { _t('State Key') } </label>
+                </div>
+                <div>
+                    <input id="stateKey" ref="stateKey" className="mx_TextInputDialog_input" size="64" />
+                </div>
+
+                <div className="mx_TextInputDialog_label">
+                    <label htmlFor="eventType"> { _t('Event Type') } </label>
+                </div>
+                <div>
+                    <input id="eventType" ref="eventType" className="mx_TextInputDialog_input" size="64" />
+                </div>
+
+                <div className="mx_TextInputDialog_label">
+                    <label htmlFor="evContent"> { _t('Event Content') } </label>
+                </div>
+                <div>
+                    <textarea id="evContent" ref="evContent" className="mx_TextInputDialog_input" defaultValue={"{\n\n}"} cols="63" rows="5" />
+                </div>
+            </div>
+            {this._buttons()}
+        </div>;
+    }
+}
+
+class RoomStateExplorer extends React.Component {
+    static propTypes = {
+        roomId: React.PropTypes.string.isRequired,
+        onBack: React.PropTypes.func.isRequired,
+    };
+
+    constructor(props, context) {
+        super(props, context);
+
+        const room = MatrixClientPeg.get().getRoom(this.props.roomId);
+        this.roomStateEvents = room.currentState.events;
+
+        this.onBack = this.onBack.bind(this);
+    }
+
+    state = {
+        eventType: null,
+        event: null,
+    };
+
+    browseEventType(eventType) {
+        const self = this;
+        return () => {
+            self.setState({ eventType });
+        };
+    }
+
+    onViewSourceClick(event) {
+        const self = this;
+        return () => {
+            self.setState({ event: event.event });
+        };
+    }
+
+    onBack() {
+        if (this.state.event) {
```

Comment:
Ordinarily I'd vote against this but given that the `setState`s are _here_, it's fairly clear what will happen when clicking "Back" several times. The alternative would be doing some routing and I don't think that's a reasonable consideration for `/devtools`.